clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -157,15 +157,23 @@ public void onMetadataUpdate(Cluster cluster) {
                 if (!cluster.unauthorizedTopics().isEmpty())
                     throw new TopicAuthorizationException(new HashSet<>(cluster.unauthorizedTopics()));
 
-                if (subscriptions.hasPatternSubscription()) {
+                boolean hasAutoTopicSubscription = subscriptions.hasAutoTopicSubscription();
+                boolean hasPatternSubscription = subscriptions.hasPatternSubscription();
+
+                if (hasAutoTopicSubscription || hasPatternSubscription) {
                     final Set<String> topicsToSubscribe = new HashSet<>();
+                    Set<String> internalTopics = cluster.internalTopics();
+                    Set<String> topics = hasAutoTopicSubscription ? subscriptions.subscription() : cluster.topics();
 
-                    for (String topic : cluster.topics())
-                        if (subscriptions.getSubscribedPattern().matcher(topic).matches() &&
-                                !(excludeInternalTopics && cluster.internalTopics().contains(topic)))
+                    for (String topic: topics)
+                        if (!(excludeInternalTopics && internalTopics.contains(topic)) &&
+                                (hasAutoTopicSubscription || subscriptions.getSubscribedPattern().matcher(topic).matches()))
                             topicsToSubscribe.add(topic);
 
-                    subscriptions.subscribeFromPattern(topicsToSubscribe);
+                    if (hasAutoTopicSubscription)
```

Comment:
I think I'm still not getting why we're trying to hook into the metadata updates in this patch. As I understand, the main point is to hold onto the current assignment metadata until the onPartitionsRevoked() hook. Unless I'm missing something, it shouldn't have anything to do with when we receive topic metadata (which can happen basically at any time).